Plugins: Update plugins list bulk management headers

johnHackworth opened the pull request:
fixes https://github.com/Automattic/wp-calypso/issues/912

This PR changes the way we bulk manage plugins in the plugins list. Until now, the plugins list header showed a "manage" button, that once clicked, it would change the view to edition mode and add a second level bar with the available actions for the selected plugins.

That secondary actions bar dissapears and the 'manage' button is now integrated in a <SectionHeader> component that is always shown on top of the differnt lists of plugins. The actions are also embed in there.

Before:
![image](https://cloud.githubusercontent.com/assets/1554855/11477993/d0193dd4-978a-11e5-963c-d2419d683a72.png)

![image](https://cloud.githubusercontent.com/assets/1554855/11477948/9fa29592-978a-11e5-91e1-54070602c963.png)

After:
![image](https://cloud.githubusercontent.com/assets/1554855/11477965/ae56455c-978a-11e5-8744-ed6d32c041ae.png)
![image](https://cloud.githubusercontent.com/assets/1554855/11477926/8810d66e-978a-11e5-9a03-a3efa0936b61.png)

johnHackworth commented:
@rickybanister @MichaelArestad It would be great if you could take a look at this. There are several points that I have not a clear image about how to proceed:

1. The amount of actions in the bar when you have several plugins selected is too large. When you are in <800px the buttons pile uglily.

2.  What should we do when the user is in "all my sites" and we have two different plugin lists? Right now I'm only showing the 'bulk edit' button in the Jetpack plugins list, because it was weird having two bulk edit buttons that did exactly the same (put both lists in edit mode and get refreshed when any of the plugins in any of the lists is selected). The idea in this case would be to have those two lists behave separatedly? like with own controls, own 'selected list', etc? because that would need a major change in most of the internal data structures.
![image](https://cloud.githubusercontent.com/assets/1554855/11478220/f62e2b96-978b-11e5-8c7f-d573fe761481.png)


3. In the previous design we had the 'bulk select' component in the top left, but now that place is occupied by the list name. Should we drop the name when we are in bulk edit mode and keep that `bulk select` component up there?

mtias commented on client/my-sites/plugins/main.jsx:
```diff
@@ -40,6 +40,8 @@ import PluginsDataStore from 'lib/plugins/wporg-data/store'
 import JetpackManageErrorPage from 'my-sites/jetpack-manage-error-page'
 import PlanNudge from 'components/plans/plan-nudge'
+import SectionHeader from 'components/section-header'
+import SectionHeaderButton from 'components/section-header/button'
```
This one should be deprecated by now in favour of just using `components/button`.

johnHackworth replied:
oops, I missed that! thanks!

mtias replied:
No pasa nada ;)

enejb commented on client/my-sites/plugins/main.jsx:
```diff
@@ -772,7 +711,7 @@ export default React.createClass( {
 					/>
 				</SectionNav>
-				{ pluginsContent }
+				{ this.getPluginsContent() }
```
Should we update this to `renderPluginsContent`?

johnHackworth replied:
yep, you're right!

johnHackworth commented:
I've added button-groups, updated the code to stop using SectionHeaderButtons and added the 'update all' button here too:

![image](https://cloud.githubusercontent.com/assets/1554855/11505355/4ebac648-984a-11e5-9cee-bff78a524685.png)

![image](https://cloud.githubusercontent.com/assets/1554855/11505357/55958cd2-984a-11e5-8560-db8b6c3a14be.png)

mtias commented:
Can we align that checkbox?

johnHackworth commented:
@mtias That's not a side effect of this PR, it has been like that forever (I thought that intendendedly). This is prod:

![image](https://cloud.githubusercontent.com/assets/1554855/11505778/6e1fc90a-984c-11e5-8fcb-06e62fb7f2a3.png)

I'll create an issue to align it

rickybanister commented:
@johnHackworth, a few answers for you:

> The amount of actions in the bar when you have several plugins selected is too large. When you are in <800px the buttons pile uglily.

We should collapse them to a dropdown, just like the section-nav if there isn't enough room.

> What should we do when the user is in "all my sites" and we have two different plugin lists? Right now I'm only showing the 'bulk edit' button in the Jetpack plugins list, because it was weird having two bulk edit buttons that did exactly the same (put both lists in edit mode and get refreshed when any of the plugins in any of the lists is selected). The idea in this case would be to have those two lists behave separatedly? like with own controls, own 'selected list', etc? because that would need a major change in most of the internal data structures.

I think we should have a button in each. I think it's ok if they both trigger bulk edit across both lists in an interconnected way. 

And finally, as you shared in a later comment, we should hide the list title when bulk editing and display the multiselect checkbox.

rickybanister commented:
A couple notes from our zoom:

- [ ] let's switch the order of things to be like so:

`activate | deactivate        autoupdate | disable autoupdates     remove`

- [ ] * `DONE` should be the search component's `X` 

- [ ] Let's show the buttons as soon as you enter bulk edit mode, but disable them until a selection has been made

- [ ] Mobile-first `SelectDropdown` component that gets hidden when there's room for all the options to be seen. Let's start by using the full-size dropdown, but if it doesn't fit we will create an `is-compact` version.

rickybanister commented:
Here's an updated design of the options:

<img width="673" alt="screen shot 2015-12-02 at 5 23 42 pm" src="https://cloud.githubusercontent.com/assets/437258/11546160/8076b88e-9919-11e5-9ab3-3d01495c2d06.png">

johnHackworth commented:
First two checks done:
![image](https://cloud.githubusercontent.com/assets/1554855/11560792/45002da8-99c3-11e5-9115-4fef612167d9.png)

![image](https://cloud.githubusercontent.com/assets/1554855/11560799/4fac1e2e-99c3-11e5-819e-af601ceec6a4.png)


(it's just me or the default 24px padding of section-header feels weird with the right side buttons / X ?)

rickybanister commented:
@johnHackworth let's do a scrub of SectionHeader tomorrow, we may be able to address the `X` in doing that. 

Currently the section-nav's search close `X` is using a noticon and is centered within a 60px span, giving it a 22px right margin. So perhaps we can get both things moved to a gridicon, and have some standardized padding (probably the same 24/16 we're using everywhere else, standardizing will likely make it feel less weird.)

johnHackworth commented:
I've added the dropdown for the mobile-resolutions, and removed the border because it looked weird:

![image](https://cloud.githubusercontent.com/assets/1554855/11590947/38f60ca4-9a94-11e5-88f9-afc062ddc075.png)

Not it looks weird when the dropdown is open:
![image](https://cloud.githubusercontent.com/assets/1554855/11590966/6e669002-9a94-11e5-83f6-91c78997c856.png)

maybe a compact version of `SelectDropdown` (which currently doesn't exist, I think) would work better

rickybanister commented:
@johnHackworth yeah, I agree, unless we make it more custom like the section-nav dropdown it seems like we'll need to find another solution.

Also having `Actions` as the default selection feels a little weird. I'd want it to not be selected by default if we need to have a default option. Can it be disabled?

johnHackworth commented:
@rickybanister sure! today it's already a little late here, but let's do a hangout next monday and go over all this stuff... funcionality-wise this is ready, we only need to figure out which styles works well (specially for the dropdown)

enejb commented:
Needs a rebase :)

johnHackworth commented:
rebased!

rickybanister commented:
@johnHackworth I talked this through with some designers. Here's what we need to do:

- [ ] Create a `compact` prop for `SelectDropdown` (possibly in another PR that gets merged before this) (maybe switch to using Gridicon in `SelectDropdown` while we're at it)
- [ ] Switch to compact `SelectDropdown` here
- [ ] Get the bulk select checkbox in here
- [ ] Add a placeholder `SectionHeader` to the plugins list loading state

johnHackworth commented:
After adding bulk-select:

![image](https://cloud.githubusercontent.com/assets/1554855/11744592/f4c2a87a-a011-11e5-95f5-729f59a646f8.png)

![image](https://cloud.githubusercontent.com/assets/1554855/11744595/ff816486-a011-11e5-92d3-b3f4aba0f372.png)

johnHackworth commented:
Added section header placeholder:
![image](https://cloud.githubusercontent.com/assets/1554855/11744914/7368fdee-a014-11e5-892e-f2c1f4413088.png)

johnHackworth commented:
updated with compact dropdowns!
![image](https://cloud.githubusercontent.com/assets/1554855/11749322/4852117c-a02d-11e5-9f0c-47e4e04b7a1b.png)

mtias commented:
Much nicer :)

MichaelArestad commented:
This is really dang nice. Huge improvement.

rickybanister commented:
I've played with it a while, there are some more things to do before bulk edit is done, but this feels ready to merge!

enejb commented:
Some notes:
- I thought we decided that we will be removing the bulk edit section from WordPress.com plugins? 
- There is width under which the checboxes don't align just right 
see 
![screen shot 2015-12-11 at 10 29 14](https://cloud.githubusercontent.com/assets/115071/11752039/2638eb46-9ff2-11e5-8f41-9ad9a9ca8e03.png)

I think the X's should be the same colour. ( one for manage the other for search )
see 
![screen shot 2015-12-11 at 10 30 33](https://cloud.githubusercontent.com/assets/115071/11752068/4b2ead14-9ff2-11e5-8374-f972deb626f4.png)

I was able to use the keyboard to navigate around and perform bulk action. It would have be nice if the x to close the manage mode would be focuses/highlighted when tabbing around. 

Other then the above minor notes the functionality is a huge improvement from the way before \o/. Great work!

enejb commented on client/components/bulk-select/docs/example.jsx:
```diff
@@ -57,7 +57,9 @@ module.exports = React.createClass( {
 				</h2>
 				<Card>
-					<BulkSelect totalElements={ this.state.elements.length } selectedElements={ this.getSelectedElementsNumber() } onToggle={ this.handleToggleAll } />
+					<div>
```
Do we need this extra div? Here what purpose does it serve?

enejb replied:
Since this is in the example code I am assuming that it is to just make it be enclosed in a block element. 
Which I think is perfectly fine. :+1:

johnHackworth replied:
yuuuuuup, exactly! I changed the bulk-select component to be a inline-block one by default, so this div is just to force a new line :)

enejb commented on client/components/select-dropdown/style.scss:
```diff
@@ -162,6 +166,14 @@
+		opacity: .5;
+	}
+
+
```
Extra new line.

enejb commented on client/my-sites/plugins/main.jsx:
```diff
@@ -41,6 +39,14 @@ import PluginNotices from 'lib/plugins/notices'
+import Gridicon from 'components/gridicon'
+import SelectDropdown from 'components/select-dropdown'
+import DropdownItem from 'components/select-dropdown/item'
+import DropdownSeparator from 'components/select-dropdown/separator'
```
I don't know if this is a new convention but are we removing the ending `;` from imports now?

johnHackworth replied:
hmmmm... I may be wrong, but I've seen it done like this in other files and I just assumed that it was already a convention... maybe not, because it's not written anywhere

rickybanister commented:
@enejb the X for search was just changed to a Gridicon—that PR just got merged, after a rebase they should be the same icon.

johnHackworth commented:
> I thought we decided that we will be removing the bulk edit section from WordPress.com plugins?

After some test and talk about it, we reberted it to having the button in both places... it looked weird with only one button when we have two lists..

enejb commented:
Sounds good. 
I also noticed the following js warnings by react. 
![screen shot 2015-12-11 at 11 01 20](https://cloud.githubusercontent.com/assets/115071/11752846/a19cc240-9ff6-11e5-945c-2c334bdcf66c.png)

johnHackworth commented:
![image](https://cloud.githubusercontent.com/assets/1554855/11778910/dd6ae6fe-a257-11e5-89e8-eac2af5d414c.png)

now buttons are aligned & are accesible by keyboard :)

johnHackworth commented:
also, all react warnings removed

rickybanister commented:
Looks like we are really close.

enejb commented:
I can still see a checkbox miss-alignment issue when the width of the page is between 480px and 660px.
See 
![screen shot 2015-12-14 at 11 27 25](https://cloud.githubusercontent.com/assets/115071/11791398/b55f9b16-a255-11e5-8276-1f0e7a34d5b7.png)

Beside the minor function renaming. Once the 2 things have been fixed up. I think this is ready to merge.

johnHackworth commented:
yup, the missalignment is a little tricky, since the margins on the search box change dinamically with width and the ones in the plugin-items are fixed (`Card` paddings). I would be inclined to make the ones in the searchbox fixed too (in another PR), but let's see what @rickybanister & @MichaelArestad have to say about it :)

johnHackworth commented:
> Beside the minor function renaming.

@enejb what functions are still to be renamed? I

enejb commented:
As per the previous discussion getPluginsContent => renderPluginContent.

johnHackworth commented:
@enejb ooooooops ouch, I would have swear that renamed that one other day...sorry! Fixed now! :)

rickybanister commented:
Regarding the breakpoint—I think it's probably best to only support jumping to 16px padding at the 480px breakpoint. There isn't a great benefit to the middle breakpoint changing the padding here—this is likely something we should see about standardizing across sections more.

The only other thing I'd really like to see come back is keeping the plugin icon visible even in bulk edit mode. I believe this might be a one-line change. 

That said, I'd be happy if we just merged this and did those two small things in a very tiny PR right after. The world needs bulk edit!

rickybanister commented:
Is there a plan for tracking bulk edit events? How many people go through with a bulk edit, which types of actions are most common, etc? I imagine they might already exist from before, but just checking.

johnHackworth commented:
@rickybanister  I agree with refactoring breakpoints in a tiny PR after this one. About the icon, it's included in this other PR : https://github.com/Automattic/wp-calypso/pull/1554

If @enejb agrees, I'll merge this tomorrow morning!

johnHackworth commented:
and about tracking, all that actions have already tracks & analytics events, so we already should have that info

johnHackworth commented:
Since it seems that everybody is agree (except for the search box paddings, which will be addressed in a different PR) and we need this merged to continue with other developments, I've rebased this and Im going to merge it
